File: 2045-second-minimum-time-to-reach-destination/2045-second-minimum-time-to-reach-destination.cpp
typedef pair<int, int> pii;
class Solution {
public:
    int secondMinimum(int n, vector<vector<int>>& edges, int time, int change) {
         unordered_map<int, vector<int>> graph;
        
        for(const auto &edge : edges){
            int u = edge[0];
            int v = edge[1];

            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        vector<int> minDistance(n+1, INT_MAX);
        vector<int> secondMinDistance(n+1, INT_MAX);

        queue<pii> pq;
        pq.push({0, 1});

        while(!pq.empty()){
            auto [timePassed, currentNode] = pq.front();
            pq.pop();

            if(currentNode == n and secondMinDistance[currentNode] != INT_MAX)
                return secondMinDistance[n];

            int currentSlot = timePassed/change;
            if(currentSlot % 2 == 1){ //currently in the red signal
                timePassed = change * (currentSlot + 1); // wait until the next region
            }

            for(const auto &adjNode : graph[currentNode]){
                if(minDistance[adjNode] > timePassed + time){
                    secondMinDistance[adjNode] = minDistance[adjNode];
                    minDistance[adjNode] =  timePassed + time;
                    pq.push({timePassed + time, adjNode});
                } else if(secondMinDistance[adjNode] > timePassed + time and 
                        minDistance[adjNode] != timePassed + time){
                      secondMinDistance[adjNode] = timePassed + time;      
                    pq.push({timePassed + time, adjNode});

                }
            }
        }

        return -1;

    }
};

/*
typedef pair<int, int> pii;
class Solution {
public:
    int secondMinimum(int n, vector<vector<int>>& edges, int time, int change) {
         unordered_map<int, vector<int>> graph;
        
        for(const auto &edge : edges){
            int u = edge[0];
            int v = edge[1];

            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        vector<int> minDistance(n+1, INT_MAX);
        vector<int> secondMinDistance(n+1, INT_MAX);

        priority_queue<pii, vector<pii>, greater<pii>> pq;
        pq.push({0, 1});

        while(!pq.empty()){
            auto [timePassed, currentNode] = pq.top();
            pq.pop();

            if(currentNode == n and secondMinDistance[currentNode] != INT_MAX)
                return secondMinDistance[n];

            int currentSlot = timePassed/change;
            if(currentSlot % 2 == 1){ //currently in the red signal
                timePassed = change * (currentSlot + 1); // wait until the next region
            }

            for(const auto &adjNode : graph[currentNode]){
                if(minDistance[adjNode] > timePassed + time){
                    secondMinDistance[adjNode] = minDistance[adjNode];
                    minDistance[adjNode] =  timePassed + time;
                    pq.push({timePassed + time, adjNode});
                } else if(secondMinDistance[adjNode] > timePassed + time and 
                        minDistance[adjNode] != timePassed + time){
                      secondMinDistance[adjNode] = timePassed + time;      
                    pq.push({timePassed + time, adjNode});

                }
            }
        }

        return -1;

    }
};
*/

File: 2045-second-minimum-time-to-reach-destination/2045-second-minimum-time-to-reach-destination.java
class Solution {
    public int secondMinimum(int n, int[][] edges, int time, int change) {
        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < n + 1; i++) {
            adjList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];

            adjList.get(u).add(v);
            adjList.get(v).add(u);
        }

        int[] minimumDistance = new int[n + 1];
        Arrays.fill(minimumDistance, Integer.MAX_VALUE);
        minimumDistance[1] = 0;

        int[] secondMinimumDistance = new int[n + 1];
        Arrays.fill(secondMinimumDistance, Integer.MAX_VALUE);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[] { 0, 1 }); // {distance, node}

        while (!pq.isEmpty()) {
            int[] curr = pq.poll();

            // If the current time/distance is even, then I am free to move to the next node.
            // else, i have to wait until the signal toggles, so that i can move. The wait time would be the below formula
            int d = (curr[0] / change) % 2 == 0 ? curr[0] : ((curr[0] / change) + 1) * change;
            int node = curr[1];

            if (node == n && secondMinimumDistance[node] != Integer.MAX_VALUE)
                return secondMinimumDistance[node];

            for (int adj : adjList.get(node)) {
                if (d + time < minimumDistance[adj]) {
                    secondMinimumDistance[adj] = minimumDistance[adj];
                    minimumDistance[adj] = d + time;
                    pq.add(new int[] { d + time, adj });
                } else if (d + time < secondMinimumDistance[adj] && d + time > minimumDistance[adj]) {
                    secondMinimumDistance[adj] = d + time;
                    pq.add(new int[] { d + time, adj });
                }
            }
        }

        return -1;
    }
}